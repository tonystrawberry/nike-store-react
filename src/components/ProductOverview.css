.product-overview{
  max-width: 592px;
  margin: 10px;
  cursor: pointer;
}

.product-photo__container{
  width: 256px;
  height: 256px;
  overflow: hidden;
}

.product-photo{
  width: 256px;
}

.product-overview__info{
  padding-top: 10px;
  padding-left: 5px;
}

.product-overview__title{
  margin: 0;
  padding: 0;
  font-weight: 300;
  font-size: 16px;
}

.product-overview__type{
  color: #acacac;
}