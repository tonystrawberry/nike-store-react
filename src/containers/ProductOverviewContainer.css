.product-overview__container{
  position: relative;
  padding: 50px;
  justify-content: center;
  margin-top: 80px;
  width: 100%;
  display: flex;
  flex: 1 1 0%;
  flex-wrap: wrap;
  height: 100vh;
}

.product-overview__container.product-selected::before{
  opacity: 0.3;
  content: '';
  cursor: pointer;
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #020202;
  transition: opacity 300ms linear;
  will-change: opacity;
}

@media (max-width: 500px){
  .product-overview__container.product-selected{
    padding: 0;
  }
}