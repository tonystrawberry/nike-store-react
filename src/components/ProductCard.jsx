import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { unselectProduct } from '../redux/actions';


import './ProductOverview.css';
import './ProductCard.css';
import Logo from '../assets/nike_logo.svg';
import NikeAirForceSmall from '../assets/nike_airforce-small.jpg';
import NikeAirForceMedium from '../assets/nike_airforce-medium.jpg';
import NikeAirForceLarge from '../assets/nike_airforce-large.jpg';


const mapStateToProps = (state) => {
  return {

  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onUnselected: () => dispatch(unselectProduct())
  }
}

class ProductCard extends PureComponent {
  constructor(props){
    super(props);

    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
    window.addEventListener('resize', this.updateDimensions);
    this.updateDimensions();
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
    window.removeEventListener('resize', this.updateDimensions);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef(node) {
    this.wrapperRef = node;
  }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.props.onUnselected();
    }
  }

  updateDimensions() {
    let x = window.matchMedia("(max-width: 1000px)");
    const photo = document.querySelectorAll('.product-card__photo');
    const pc = document.querySelector('.product-card__photo-container');
  
    if (x.matches){
      // shoeBackgroundの新しいheight計算
      let width = photo[0].offsetWidth; // offsetHeight = elementのheight + elementのpadding + elementのborder
      pc.style.height = `${width}px`;
    } else {
      pc.style.height = '475px';
    }
  }

  render() {
    return (
      <div ref={this.setWrapperRef} className="product-card__container">
        <div className="product-card__card">
          <div className="product-card__photo-container">
            <img src={Logo} alt="Nike Logo" className="product-card__logo" />
            <a href="#" className="product-card__share"><i className="fas fa-share-alt"></i></a>
            <img src={NikeAirForceLarge} 
              srcSet={`${NikeAirForceSmall} 216w, ${NikeAirForceMedium} 432w, ${NikeAirForceLarge} 864w`} 
              sizes='(max-width: 500px) 432px, 864px'
              alt="Product Photography" className="product-card__photo"/>
          </div>
          <div className="product-card__info">
            <div className="product-card__name">
              <div>
                <h1 className="big">Nike Air Force 1</h1>
                {this.props.new ? <span className="new">new</span> : null}
              </div>
              <h3 className="small">Men's Shoe</h3>
            </div>
            <div className="product-card__description">
              <p className="product-card__text">The legend lives on in the Nike Air Force 1, which stays true to its roots with iconic AF1 style and Nike Air for all-day comfort and long-lasting wear.</p>
            </div>
            <div className="product-card__color-container">
              <h3 className="product-card__title">Color</h3>
              <div className="product-card__colors">
                <span className="product-card__color" primary="#2175f5" color="blue"></span>
                <span className="product-card__color" primary="#f84848" color="red"></span>
                <span className="product-card__color" primary="#29b864" color="green"></span>
                <span className="product-card__color" primary="#ff5521" color="orange"></span>
                <span className="product-card__color" primary="#444" color="black"></span>
              </div>
            </div>
            <div className="product-card__size-container">
              <h3 className="product-card__title">size</h3>
              <div className="product-card__sizes">
                <span className="product-card__size">7</span>
                <span className="product-card__size">8</span>
                <span className="product-card__size">9</span>
                <span className="product-card__size">10</span>
                <span className="product-card__size">11</span>
              </div>
            </div>
            <div className="product-card__buy-price">
              <a href="#" className="product-card__buy"><i className="fas fa-shopping-cart"></i>Add to cart</a>
              <div className="product-card__price">
                <i className="fas fa-yen-sign"></i>
                <h1>21,000</h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);