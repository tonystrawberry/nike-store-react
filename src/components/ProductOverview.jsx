import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { selectProduct } from '../redux/actions';

import NikeAirForceSmall from '../assets/nike_airforce-small.jpg';

import './ProductOverview.css';

const mapStateToProps = (state) => {
  return {

  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onSelected: (productId) => dispatch(selectProduct(productId))
  }
}

class ProductOverview extends PureComponent {
  render() {
    return (
      <div className="product-overview" onClick={() => this.props.onSelected(1)}>
        <div className="product-photo__container">
          <img className="product-photo" src={NikeAirForceSmall} srcSet={`${NikeAirForceSmall} 216w`} alt="Nike Air Force"></img>
        </div>
        <div className="product-overview__info">
          <h1 className="product-overview__title">Nike Air Force 1</h1>
          <div className="product-overview__type">Men's Shoe</div>
          <div className="product-overview__type">2 Colors</div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductOverview);