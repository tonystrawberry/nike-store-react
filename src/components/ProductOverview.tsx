import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { selectProduct } from '../redux/actions';
import { Dispatch } from 'redux';

import './ProductOverview.scss';
import { Product } from '../types';

const mapStateToProps = () => {
  return {

  }
}

interface IProductOverviewProps {
  onSelected: {( productId : string): void},
  product: Product
}

interface IProductOverviewState {
  
}

const mapDispatchToProps = (dispatch : Dispatch) => {
  return {
    onSelected: (productId : string) => dispatch(selectProduct(productId))
  }
}

class ProductOverview extends PureComponent<IProductOverviewProps, IProductOverviewState> {
  render() {
    return (
      <div className="product-overview" onClick={() => this.props.onSelected(this.props.product._id)}>
        <div className="product-photo__container">
          <img className="product-photo" src={this.props.product.imageUrl} alt="Nike Air Force"></img>
        </div>
        <div className="product-overview__info">
          <h1 className="product-overview__title">{this.props.product.title}</h1>
          <div className="product-overview__type">{this.props.product.subtitle1}</div>
          <div className="product-overview__type">{this.props.product.subtitle2}</div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductOverview);