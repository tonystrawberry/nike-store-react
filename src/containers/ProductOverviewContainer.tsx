import React, { PureComponent, Suspense } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import ProductOverview from '../components/ProductOverview';
import Loading from '../components/Loading';

import { ProductOverviewState, Product } from '../types';

import './ProductOverviewContainer.scss';

const ProductCardLazy = React.lazy(() => import('../components/ProductCard'));

const mapStateToProps = (state: { productOverview : ProductOverviewState }) => {
  return {
    selected: state.productOverview.selected,
    loading: state.productOverview.loading,
    products: state.productOverview.products,
    selectedProductId: state.productOverview.selectedProductId
  }
}

const mapDispatchToProps = (dispatch : Dispatch ) => {
  return {
  }
}
interface IProductOverviewContainerProps {
  selected: boolean,
  loading: boolean,
  products: Product[],
  selectedProductId: string | null
}

interface IProductOverviewContainerState {
  
}

class ProductOverviewContainer extends PureComponent<IProductOverviewContainerProps, IProductOverviewContainerState> {
  
  render() {

    if (this.props.loading){
      return(
        <div className="product-overview__container">
          <Loading />
        </div>
      )
    }

    var product = this.props.products.find(product => product._id === this.props.selectedProductId)

    return (
      <div className={`product-overview__container ${this.props.selected ? 'product-selected' : ''}`}>
        {!this.props.selected ?
          <React.Fragment>
            {this.props.products.map((product: Product) => {
              return <ProductOverview key={product._id} product={product}/>
            })}
          </React.Fragment>
          : ( this.props.selectedProductId != null ? 
            <Suspense fallback={<Loading />}>
              <ProductCardLazy new={false} product={product}/>
            </Suspense>
            :
            <Loading />
          )
        }
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductOverviewContainer);