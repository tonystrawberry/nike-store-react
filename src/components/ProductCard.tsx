import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { unselectProduct } from '../redux/actions';
import { Dispatch } from 'redux';

import './ProductOverview.scss';
import './ProductCard.scss';
import Logo from '../assets/nike_logo.svg';
import { Product } from '../types';
import Loading from './Loading';
import { numberToPrice } from '../utils/utils';


const mapStateToProps = () => {
  return {

  }
}

const mapDispatchToProps = (dispatch : Dispatch ) => {
  return {
    onUnselected: () => dispatch(unselectProduct())
  }
}

interface IProductCardProps {
  new: boolean,
  onUnselected: {(): void},
  product: Product | undefined
}

interface IProductCardState {
  
}


class ProductCard extends PureComponent<IProductCardProps, IProductCardState> {
  constructor(props : IProductCardProps){
    super(props);

    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.wrapperRef = null;
  }

  wrapperRef: HTMLDivElement | null;

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
    window.addEventListener('resize', this.updateDimensions);
    this.updateDimensions();
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
    window.removeEventListener('resize', this.updateDimensions);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef(node : HTMLDivElement | null) : void{
    this.wrapperRef= node;
  }

  /**
   * Alert if clicked on outside of element
   */
  handleClickOutside(event : Event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target as Node)) {
      this.props.onUnselected();
    }
  }

  updateDimensions() {
    let x = window.matchMedia("(max-width: 1000px)");
    const photo = document.querySelectorAll('.product-card__photo') as NodeListOf<HTMLElement>;
    const pc = document.querySelector('.product-card__photo-container') as HTMLElement;
  
    if (x.matches){
      // shoeBackgroundの新しいheight計算
      let width = photo[0].offsetWidth; // offsetHeight = elementのheight + elementのpadding + elementのborder
      pc.style.height = `${width}px`;
    } else {
      pc.style.height = '475px';
    }
  }

  render() {
    if (!this.props.product){
      return (
        <div className="product-card__container">
          <Loading />
        </div>)
    }
    return (
      <div ref={this.setWrapperRef} className="product-card__container">
        <div className="product-card__card">
          <div className="product-card__photo-container">
            <img src={Logo} alt="Nike Logo" className="product-card__logo" />
            <a href="/" className="product-card__share"><i className="fas fa-share-alt"></i></a>
            <img src={this.props.product.imageUrl} 
              alt="Product Photography" className="product-card__photo"/>
          </div>
          <div className="product-card__info">
            <div className="product-card__name">
              <div>
                <h1 className="big">{ this.props.product.title }</h1>
                {this.props.new ? <span className="new">new</span> : null}
              </div>
              <h3 className="small">{ this.props.product.subtitle1 }</h3>
              <h3 className="small">{ this.props.product.subtitle2 }</h3>
            </div>
            <div className="product-card__description">
              <p className="product-card__text">{ this.props.product.description }</p>
            </div>
            {/* <div className="product-card__color-container">
              <h3 className="product-card__title">Color</h3>
              <div className="product-card__colors">
                <span className="product-card__color" color="blue"></span>
                <span className="product-card__color" color="red"></span>
                <span className="product-card__color" color="green"></span>
                <span className="product-card__color" color="orange"></span>
                <span className="product-card__color" color="black"></span>
              </div>
            </div>
            <div className="product-card__size-container">
              <h3 className="product-card__title">size</h3>
              <div className="product-card__sizes">
                <span className="product-card__size">7</span>
                <span className="product-card__size">8</span>
                <span className="product-card__size">9</span>
                <span className="product-card__size">10</span>
                <span className="product-card__size">11</span>
              </div>
            </div> */}
            <div className="product-card__buy-price">
              <a href="/" className="product-card__buy"><i className="fas fa-shopping-cart"></i>Add to cart</a>
              <div className="product-card__price">
                <i className="fas fa-yen-sign"></i>
                <h1>{ numberToPrice(this.props.product.price) }</h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);