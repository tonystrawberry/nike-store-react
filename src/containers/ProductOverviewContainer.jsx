import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import ProductOverview from '../components/ProductOverview';
import ProductCard from '../components/ProductCard';

import './ProductOverviewContainer.css';

const mapStateToProps = (state) => {
  return {
    selected: state.selectProduct.selected,
    selectedProductId: state.selectProduct.selectedProductId
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    
  }
}

class ProductOverviewList extends PureComponent {

  render() {
    return (
      <div className={`product-overview__container ${this.props.selected ? 'product-selected' : ''}`}>
        {!this.props.selected ?
          <React.Fragment>
            <ProductOverview />
            <ProductOverview />
            <ProductOverview />
            <ProductOverview />
            <ProductOverview />
            <ProductOverview />
          </React.Fragment>
          :
          <ProductCard />
        }
      </div>
    );
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductOverviewList);