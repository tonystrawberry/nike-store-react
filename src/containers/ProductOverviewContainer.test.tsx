import { mount, shallow } from 'enzyme';
import React, { Suspense } from 'react';
import configureStore from 'redux-mock-store';
import ProductOverviewContainer from './ProductOverviewContainer';
import ProductOverview from '../components/ProductOverview';
import ProductCard from '../components/ProductCard';
import toJson from 'enzyme-to-json';

import { Provider } from 'react-redux';

var initialStateSelectedFalse = {
  selectProduct: {
    selected: false,
    selectedProductId: null,
    products: [],
    loading: false
  }
}

var initialStateSelectedTrue = {
  selectProduct: {
    selected: true,
    selectedProductId: 1,
    products: [],
    loading: false
  }
}

const mockStore = configureStore([]);
let wrapper;
let store;

beforeEach(() => {
})

it ('expect to render 6 ProductOverview components', () => {
  store = mockStore(initialStateSelectedFalse);
  wrapper = mount(<Provider store={store}><ProductOverviewContainer /></Provider>);

  expect(wrapper.find(ProductOverview).length).toEqual(0)
  expect(wrapper.find(ProductCard).length).toEqual(0)

})


it ('expect to render 1 ProductCartd component', async() => {
  store = mockStore(initialStateSelectedTrue);
  wrapper = mount(<Provider store={store}>
      <ProductOverviewContainer />
  </Provider>);

  expect(wrapper.find(ProductOverview).length).toEqual(0)
  expect(toJson(wrapper)).toMatchSnapshot()
})




