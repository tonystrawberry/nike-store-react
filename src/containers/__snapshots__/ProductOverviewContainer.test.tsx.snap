// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render 1 ProductCart component on selected: true 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ProductOverviewContainer)>
    <ProductOverviewContainer
      fetchProducts={[Function]}
      fetchProductsError={[Function]}
      fetchProductsSuccess={[Function]}
      loading={false}
      products={Array []}
      selected={true}
    >
      <div
        className="product-overview__container product-selected"
      >
        <Suspense
          fallback={<Loading />}
        >
          <Loading>
            <div>
              <img
                alt="Loading"
                src="loading.gif"
                width="50"
              />
            </div>
          </Loading>
        </Suspense>
      </div>
    </ProductOverviewContainer>
  </Connect(ProductOverviewContainer)>
</Provider>
`;
